xio-core/src/main/java/com/xjeffrose/xio/http/Client.java

```diff
@@ -78,39 +50,103 @@ public ChannelFuture connect() {
+      // If we are not in a connected state we should buffer the requests until we find out
+      // what happened to the connection try.  The connectFuture calls back on the same eventloop,
+      // since we are never reconnecting clients for different serverchannel eventloops
+      log.debug(
+          "== No channel exists, lets connect on client: " + this + " with request: " + request);
+      ChannelFuture connectFuture = manager.connect();
+      promise = manager.currentChannel().newPromise();
+      log.debug("== Adding req: " + request + " to queue on client: " + this);
+      this.requestQueue.add(new Client.ClientPayload(request, promise));
+      connectFuture.addListener(
+          (connectionResult) -> {
+            executeBufferedRequests(connectionResult);
+          });
+      return Optional.of(promise);
+    } else if (manager.connectionState() == ClientConnectionState.CONNECTING) {
+      // we are in the middle of connecting so lets just add to the queue
+      // this is a non concurrent queue because these write calls methods will be called on the
+      // same eventloop as the connectFuture.listener callback. We do not reconnect on previously
+      // connected clients so we don't have to worry about new serverchannel's trying to call connect
+      // on a client that was bound to previous serverchannel's eventloop
+      promise = manager.currentChannel().newPromise();
+      log.debug("== Adding req: " + request + " to queue on client: " + this);
+      this.requestQueue.add(new Client.ClientPayload(request, promise));
+      return Optional.of(promise);
+    } else if (manager.connectionState() == ClientConnectionState.CONNECTED) {
+      // we are already connected so fire away
+      log.debug("== already connected, just writing req: " + request + " on client: " + this);
+      return Optional.of(this.rawWrite(request));
     } else {
-      if (request.endOfMessage()) {
-        return channel.writeAndFlush(request).addListener(writeListener);
+      log.debug("== Connect failed on client: " + this);
+      return Optional.empty();
+    }
+  }
+
+  private ChannelFuture rawWrite(Request request) {
+    return request.endOfMessage()
+        ? manager.currentChannel().writeAndFlush(request).addListener(this.writeListener)
+        : manager.currentChannel().write(request).addListener(this.writeListener);
+  }
+
+  private void executeBufferedRequests(Future<? super Void> connectionResult) {
+    boolean connectionSuccess = connectionResult.isDone() && connectionResult.isSuccess();
+    log.debug("== Connection success was " + connectionSuccess);
+    // loop through the queue until it's empty and fire away
+    // this will happen on the same eventloop as the write so we don't need to worry about
+    // trying to write to this queue at the same time we are dequeing
+    while (!requestQueue.isEmpty()) {
+      Client.ClientPayload pair = (Client.ClientPayload) requestQueue.remove();
```

Comment:
Is `pair` a normal word to describe a `ClientPayload`?

Reply:
I can rename it, i really just wish we had tuple =p

Reply:
ok updated it to requestPayload instead of pair